kramer/src/main/kotlin/kramer/MavenRepo.kt

```diff
@@ -0,0 +1,331 @@
+    "--workspace",
+    help = "Path to the workspace to be generated."
+  )
+    .path(canBeFile = false, canBeSymlink = false, canBeDir = true)
+    .default(fs.getPath("${System.getProperties()["java.io.tmpdir"]}", "bazel/maven"))
+
+  /** Name of the workspace - nearly always the nearest directory name of the workspace path */
+  private val workspaceName: String by option("--workspace-name")
+    .defaultLazy { workspace.fileName.toString() }
+
+  private val configFile by option("--configuration").path(mustExist = true).required()
+
+  private val threadCount: Int by option("--threads").int().default(1)
+
+  internal val kontext by requireObject<Kontext>()
+
+  internal data class IndexEntry(
+    val dependants: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap()),
+    val versions: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap())
+  )
+
+  @FlowPreview
+  @ExperimentalCoroutinesApi
+  override fun run() {
+    val repoConfig = parseRepoConfig(configFile)
+    val count = AtomicInteger(0)
+    var exit = 0
+    val seen = ConcurrentHashMap<String, IndexEntry>()
+    val unresolved = Collections.newSetFromMap(ConcurrentHashMap<String, Boolean>())
+    val declaredArtifactSlugs = Collections.newSetFromMap(ConcurrentHashMap<String, Boolean>())
+    val issue = ConcurrentHashMap<String, String>()
+    val benchmark = measureTimeMillis {
+      if (repoConfig.artifacts.isNotEmpty()) Files.createDirectories(workspace)
+      runBlocking {
+        kontext.out(false) { "Building workspace for ${repoConfig.artifacts.size} artifacts" }
+        kontext.out { if (kontext.verbosity > 1) "using $threadCount threads." else "." }
+        repoConfig.artifacts.entries.asFlow()
+          .onEach {
+            // TODO: Validate per-artifact configuration.
+          }
+          .flatMapMerge(concurrency = threadCount) { (artifact_spec, config) ->
+            flow {
+              val resolver = ArtifactResolver(
+                cacheDir = kontext.localRepository,
+                suppressAddRepositoryWarnings = true,
+                repositories =
+                if (kontext.repositories.isNotEmpty()) kontext.repositories
+                else Repositories.DEFAULT
+              )
+              val artifact = resolver.artifactFor(artifact_spec)
+              val slug = "${artifact.groupId}:${artifact.artifactId}"
+              val entry = seen.getOrPut(slug) { IndexEntry() }
+              entry.versions.add(artifact.version)
+              declaredArtifactSlugs.add(slug)
+              var resolved: ResolvedArtifact? = null
+              val time = measureTimeMillis {
+                resolved = resolver.resolveArtifact(artifact)
+              }
+              val success = if (resolved != null) "successfully" else "unsuccessfully"
+              kontext.info { "Resolved ${artifact.coordinate} $success in ${time / 1000.0} seconds" }
+              resolved?.model?.apply {
```

Comment:
Is this a real null check or just to work around the measureTimeMillis?

If it just for millis, I'd suggest adding a profile function to kontext. Flows better than null checking.

Reply:
It's real. If maven-architect can't resolve it it ends up being null, and it only makes sense to emit resolved in to the flow if it was, in fact, resolved.

Reply:
Yeah, it would be clearer to handle the error condition and fail/return early.

Reply:
I can't do that and still gather all the errors together. The biggest early complaint about this software is that it was failing early, and thereby masking other similar or related errors. But I think I made it more obvious - moved the logic from here and the gathering logic in teh subsequent filter up into an if/else, which should be more obvious.